src/ripple/protocol/QualityFunction.h

```diff
@@ -0,0 +1,107 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2023 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_PROTOCOL_QUALITYFUNCTION_H_INCLUDED
+#define RIPPLE_PROTOCOL_QUALITYFUNCTION_H_INCLUDED
+
+#include <ripple/basics/Number.h>
+
+namespace ripple {
+
+class Quality;
+
+/** Average quality of a path as a function of `out`: q(out) = m * out + b,
+ * where m = -1 / poolGets, b = poolPays / poolGets. If CLOB offer then
+ * `m` is equal to 0 `b` is equal to the offer's quality. The function
+ * is derived by substituting `in` in q = out / in with the swap out formula
+ * for `in`:
+ * in = [(poolGets * poolPays) / (poolGets - out) - poolPays] / (1 - tfee)
+ * and combining the function for multiple steps. The function is used
+ * to limit required output amount when quality limit is provided in one
+ * path optimization.
+ */
+class QualityFunction
+{
+private:
+    // slope
+    Number m_;
+    // intercept
+    Number b_;
+    // seated if QF is for CLOB offer.
+    std::optional<Quality> quality_;
+
+public:
+    struct AMMTag
+    {
+    };
+    // AMMOffer for multi-path is like CLOB, i.e. the offer size
+    // changes proportionally to its quality.
+    struct CLOBLikeTag
+    {
+    };
+    QualityFunction(Quality const& quality, CLOBLikeTag);
+    template <typename TIn, typename TOut>
+    QualityFunction(
+        TAmounts<TIn, TOut> const& amounts,
+        std::uint32_t tfee,
+        AMMTag);
+
+    /** Combines QF with the next step QF
+     */
+    void
+    combineWithNext(QualityFunction const& qf);
```

Comment:
Nit: I think this function should just be named `combine`. The `QualityFunction` doesn't care what it's being combined with, and there is nothing like `combineWithPrevious` to distinguish it from.

Reply:
It cares that it must be combined with the next step, otherwise the calculations are going to be invalid. It doesn't matter if the steps are BookSteps with account offers because the quality is constant in this case. But it does matter when the steps are AMM offers because the quality is a linear function.

Reply:
At the risk of bike shedding, my point is that the `QualityFunction` doesn't "know" what it's being combined with. The caller is responsible for "knowing" what it's combining.

If you still disagree, though, I won't argue further.

Reply:
Will update.